File: export_decomp.py
# Ghidra headless script to export decompiled C code
# @category: Export

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
import os

decomp = DecompInterface()
decomp.openProgram(currentProgram)

monitor = ConsoleTaskMonitor()

output_dir = os.path.join(os.path.expanduser("~"), "anno", "decompiled")
if not os.path.exists(output_dir):
    os.makedirs(output_dir)

prog_name = currentProgram.getName().replace(".", "_")
output_file = os.path.join(output_dir, prog_name + ".c")

listing = currentProgram.getListing()
func_manager = currentProgram.getFunctionManager()

with open(output_file, "w") as f:
    f.write("// Decompiled from: %s\n\n" % currentProgram.getName())

    # Export all functions
    func_iter = func_manager.getFunctions(True)
    count = 0
    for func in func_iter:
        if monitor.isCancelled():
            break
        results = decomp.decompileFunction(func, 30, monitor)
        if results and results.decompileCompleted():
            decomp_func = results.getDecompiledFunction()
            if decomp_func:
                sig = decomp_func.getSignature()
                body = decomp_func.getC()
                f.write("// Function at 0x%s\n" % func.getEntryPoint().toString())
                f.write(body)
                f.write("\n\n")
                count += 1

    print("Exported %d functions to %s" % (count, output_file))

decomp.dispose()

File: ExportDecomp.java
// Ghidra script to export decompiled C code
// @category Export
// @author Anno1602 RE

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;

public class ExportDecomp extends GhidraScript {
    @Override
    public void run() throws Exception {
        DecompInterface decomp = new DecompInterface();
        decomp.openProgram(currentProgram);

        String homeDir = System.getProperty("user.home");
        File outputDir = new File(homeDir + "/anno/decompiled");
        outputDir.mkdirs();

        String progName = currentProgram.getName().replace(".", "_");
        File outputFile = new File(outputDir, progName + ".c");

        PrintWriter writer = new PrintWriter(new FileWriter(outputFile));
        writer.println("// Decompiled from: " + currentProgram.getName());
        writer.println();

        FunctionIterator funcIter = currentProgram.getFunctionManager().getFunctions(true);
        int count = 0;
        while (funcIter.hasNext() && !monitor.isCancelled()) {
            Function func = funcIter.next();
            DecompileResults results = decomp.decompileFunction(func, 30, monitor);
            if (results != null && results.decompileCompleted()) {
                if (results.getDecompiledFunction() != null) {
                    String body = results.getDecompiledFunction().getC();
                    writer.println("// Function at " + func.getEntryPoint().toString());
                    writer.println("// Name: " + func.getName());
                    writer.println(body);
                    writer.println();
                    count++;
                }
            }
        }
        writer.close();
        decomp.dispose();
        println("Exported " + count + " functions to " + outputFile.getAbsolutePath());
    }
}
